### PR TITLE
DOC: add details on default build options used by meson-python

### PR DESCRIPTION
**Pull request by rgommers**
This follows up on the `n_debug` default added in gh-325, documents all other default options and they are used, as well as improves the page on using `pyproject.toml` settings.

### CODE REVIEW

File: docs/explanations/default-options.rst
Comment (dnicolodi) [inline docs/explanations/default-options.rst:13]: Forced?

File: docs/explanations/default-options.rst
Comment (dnicolodi) [inline docs/explanations/default-options.rst:38]: These are going away, see #279.

File: docs/reference/pyproject-settings.rst
Comment (dnicolodi) [inline docs/reference/pyproject-settings.rst:49]: This was fixed in #304. In general, I think it is a good idea to link to a bug report any time we say "in a future version" so that the user can check whether that means in practice.

File: docs/explanations/default-options.rst
Comment (dnicolodi) [inline docs/explanations/default-options.rst:32]: The native file will not named like this anymore in the next meson-python release. It will be in the build directory and names `meson-python-native-file.ini`. One thing that is not obvious is that multiple `--native-file` options can be passed and Meson will merge the contents. Therefore, if the user needs to pass any other configuration via a native file, they can add another `--native-file` argument, as long as they do not try to set the path for the `python` executable. I think it would be useful to mention this in the documentation.

File: docs/reference/pyproject-settings.rst
Comment (rgommers) [reply docs/reference/pyproject-settings.rst:49]: Ah, great!

>  In general, I think it is a good idea to link to a bug report any time we say "in a future version" so that the user can check whether that means in practice.

Agreed. I looked for one but couldn't find it. I think it was discussed in other PR review.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:38]: That's merged, but these options are still there. They're not specific to editable installs, so I'm not sure I understand what you meant here.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:13]: I'll rephrase as "needs to" if that's better. Overrides for, e.g., `--prefix` are necessary.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:32]: Agreed, thanks. Will update and add that info.

File: docs/reference/pyproject-settings.rst
Comment (rgommers) [reply docs/reference/pyproject-settings.rst]: Removed this note, so resolving the comment.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst]: added now

File: docs/explanations/default-options.rst
Comment (FFY00) [reply docs/explanations/default-options.rst:42]: I think Daniele meant #340. I'll review that tomorrow, but we can either wait until that gets a resolution, or merge this and have that PR update it.

File: docs/explanations/default-options.rst
Comment (dnicolodi) [reply docs/explanations/default-options.rst]: Technically, they are not necessary. We could very well use the default prefix and use the `$DESTDIR` environment variable to install in a temporary directory. I would even argue that this would be better. Anyhow, it does not matter.

File: docs/explanations/default-options.rst
Comment (dnicolodi) [reply docs/explanations/default-options.rst:42]: Yes, sorry, I linked to the wrong PR.

File: docs/explanations/default-options.rst
Comment (FFY00) [reply docs/explanations/default-options.rst]: Yes, I think it was only needed for the Debian heuristics, I think we can drop it now that Meson uses directory placeholders for all paths.

File: docs/explanations/default-options.rst
Comment (FFY00) [reply docs/explanations/default-options.rst:42]: Because of #341 and #344, we cannot get rid of the these options at the moment, so they should stay for now.

File: docs/explanations/default-options.rst
Comment (FFY00) [inline docs/explanations/default-options.rst:14]: possible improvement:

We could split the paragraph here and also document why we might choose to override the defaults. I wrote about the reasoning in https://github.com/mesonbuild/meson-python/pull/325#issuecomment-1463902837.

We could also note that the options we do need to set are mostly for build/installation details that should not be relevant to the user. The only exception is the native file, but the user should be able to override the parts of it that make sense (IIRC everything but the Python interpreter).

File: docs/explanations/default-options.rst
Comment (FFY00) [inline docs/explanations/default-options.rst:28]: possible improvement:

I'd group the options we need to set, with the ones we choose to override the default. Something like this:

```suggestion
      (we need override)
      prefix           : /home/username/mambaforge/envs/project-dev
      python.platlibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
      python.purelibdir: /home/username/mambaforge/envs/project-dev/lib/python3.10/site-packages
      Native files     : /home/username/code/project/.mesonpy-native-file.ini
      (we chose to override)
      debug            : false
      optimization     : 2
      b_ndebug         : if-release
```

It might make sense to make the distinction, so that users know what they can and can't override.

File: docs/explanations/default-options.rst
Comment (FFY00) [inline docs/explanations/default-options.rst:50]:
```suggestion
.. code-block:: console
```

This makes it so that the $ is not selectable, excluding it from ctrl+a or the copy button from the sphinx extension.

File: docs/explanations/default-options.rst
Comment (FFY00) [inline docs/explanations/default-options.rst:64]: nitpick

```suggestion
For more details on overriding build options at build time see the
:ref:`reference-config-settings` page, and in ``pyproject.toml`` see
the :ref:`reference-pyproject-settings` page.
```

File: docs/reference/pyproject-settings.rst
Comment (FFY00) [inline docs/reference/pyproject-settings.rst:35]: possible improvement:

I think this may be better as sub-section.

```suggestion
Example
=====
```

**Review by FFY00 — APPROVED** [review]
@rgommers thanks! I have a one comment, and a couple of suggestions, but overall looks good.

File: docs/explanations/default-options.rst
Comment (dnicolodi) [reply docs/explanations/default-options.rst:28]: This is the output of the `meson setup` command. Presenting an edited version of it that does not match what the uses sees would be confusing.

File: docs/explanations/default-options.rst
Comment (dnicolodi) [reply docs/explanations/default-options.rst:14]: If we keep `meson install` I would much prefer if we would get rid of the `--prefix` and `--python.{plat,pure}lib` options and use the `$DESTDIR` environment variable pointing to somewhere in `/tmp` instead. This would have the additional benefit of putting the installation directory on tmpfs, with possible significant speedups when building large wheels.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:28]: Agreed, I just Ctrl-C'd this from actual build output (replacing my own user name with `username` and package name with `project`) and would prefer to leave it as is.

File: docs/reference/pyproject-settings.rst
Comment (rgommers) [reply docs/reference/pyproject-settings.rst:35]: This was on purpose - subsection heading look terrible in our Sphinx theme, and we should not over-use them.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:14]: > We could split the paragraph here and also document why we might choose to override the defaults. I wrote about the reasoning in [#325 (comment)](https://github.com/mesonbuild/meson-python/pull/325#issuecomment-1463902837).

I checked it in more detail, and it's effectively the same as what I already have in this PR: _"because Meson defaults to values that are appropriate for development, while the main purpose of meson-python is to build release artifacts."_

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst]: adopted, resolving

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst]: thanks, adopted

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:28]: rejecting the suggested change - resolving.

**Comment by dnicolodi** [comment]
It would be nice to have this combined somehow with the "Passing arguments to Meson" section. By the way, I just read that section and it contain several inaccuracies and some errors, for example all the pip command lines are wrong.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:42]: The current text is good as is given the current state of the repo, I think we can resolve this comment.

**Comment by rgommers** [comment]
> It would be nice to have this combined somehow with the "Passing arguments to Meson" section.

**Comment by rgommers** [comment]
> It would be nice to have this combined somehow with the "Passing arguments to Meson" section.

Ah, that is new. I don't have more time to spend on this right now, maybe end of the week. It looks to me like only the end of the page overlaps. How about leaving the explanation here, and the two examples of overriding arguments can be merged into the other page (so everything after _"It is possible to override these defaults, either permanently in your project or at build time."_)?

File: docs/explanations/default-options.rst
Comment (FFY00) [reply docs/explanations/default-options.rst:14]: > If we keep `meson install` I would much prefer if we would get rid of the `--prefix` and `--python.{plat,pure}lib` options and use the `$DESTDIR` environment variable pointing to somewhere in `/tmp` instead.

I don't think this is even a question, is it :rofl:? Those options and `DESTDIR` serve two different purposes. We only set those options for the heuristics, so they can be removed now.

Regarding `DESTDIR`, we already set it to a subdirectory of `working_dir`, which currently when using the PEP 517 hooks is always set by `tempfile.TemporaryDirectory` (in `with_temp_working_dir`), so isn't that true already?

Comment (dnicolodi) [reply docs/explanations/default-options.rst:14]: The working dir is created inside the project directory, thus it is on the same filesystem as the project sources.

**Comment by dnicolodi** [comment]
Seems appropriate. I just noticed that you placed this page in the "Explanations" section. Why? I think it would be better in the "Reference" section. Anyhow, I think we can merge it as it is and tweak it up later.

File: docs/explanations/default-options.rst
Comment (FFY00) [reply docs/explanations/default-options.rst:14]: Ah, you're right :sweat_smile:. IIRC I did that to make things work properly for packagers by default, where access to `/tmp` might be a bit tricky. We can consider changing the default, though.

File: docs/explanations/default-options.rst
Comment (FFY00) [inline docs/explanations/default-options.rst:60]: Since #291 is not merged, wouldn't it be better to link to that page?

https://meson-python.readthedocs.io/en/latest/how-to-guides/meson-args.html

**Comment by FFY00** [comment]
I think the page as-is should be in the explanations section, as its main purpose is to explain the reason behind the options we set, but it might make sense to have a page in the reference section detailing the default options we override.

**Comment by FFY00** [comment]
@rgommers feel free to apply any of the changes suggested, if you want, and then merge.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst]: Agreed, done.

File: docs/explanations/default-options.rst
Comment (rgommers) [reply docs/explanations/default-options.rst:14]: I'll leave this comment thread open, because it seems like useful info. It doesn't apply to this PR directly though; it's discussion of a potential future design change. If that materializes, the docs can simply be updated.

**Comment by rgommers** [comment]
The pre-commit failure is unrelated to this PR, it complains about a missing `stacklevel`. I'll ignore here.

**Pull request merged**
